frame/staking/src/pallet/impls.rs

```diff
@@ -754,65 +811,91 @@ impl<T: Config> Pallet<T> {
 	/// Get the targets for an upcoming npos election.
 	///
 	/// This function is self-weighing as [`DispatchClass::Mandatory`].
-	pub fn get_npos_targets() -> Vec<T::AccountId> {
-		let mut validator_count = 0u32;
-		let targets = Validators::<T>::iter()
-			.map(|(v, _)| {
-				validator_count.saturating_inc();
-				v
-			})
+	pub fn get_npos_targets(maybe_max_len: Option<usize>) -> Vec<T::AccountId> {
+		let targets = T::TargetList::iter()
+			.take(maybe_max_len.unwrap_or_else(Bounded::max_value))
 			.collect::<Vec<_>>();
 
-		Self::register_weight(T::WeightInfo::get_npos_targets(validator_count));
+		Self::register_weight(T::WeightInfo::get_npos_targets(targets.len() as u32));
+		log!(info, "generated {} npos targets", targets.len());
 
 		targets
 	}
 
 	/// This function will add a nominator to the `Nominators` storage map,
-	/// and `VoterList`.
+	/// and [`Config::VoterList`].
 	///
 	/// If the nominator already exists, their nominations will be updated.
 	///
 	/// NOTE: you must ALWAYS use this function to add nominator or update their targets. Any access
 	/// to `Nominators` or `VoterList` outside of this function is almost certainly
 	/// wrong.
-	pub fn do_add_nominator(who: &T::AccountId, nominations: Nominations<T>) {
-		if !Nominators::<T>::contains_key(who) {
+	pub fn do_add_nominator(
+		stash: &T::AccountId,
+		old: Vec<T::AccountId>,
+		nominations: Nominations<T>,
+	) {
+		if !NominatorsHelper::<T>::contains_any(stash) {
 			// maybe update sorted list.
-			let _ = T::VoterList::on_insert(who.clone(), Self::weight_of(who))
-				.defensive_unwrap_or_default();
+			let _ = T::VoterList::on_insert(stash.clone(), Self::weight_of(stash)).defensive();
```

Comment:
Just noting that we must ensure the account is removed as a validator prior to being added as a nominator. otherwise this would be a second insert of an account that already was added to the list as a nominator